direct/core/src/main/java/cz/o2/proxima/direct/transaction/TransactionalOnlineAttributeWriter.java

```diff
@@ -0,0 +1,215 @@
+          break;
+        default:
+          throw new TransactionRejectedException(transactionId);
+      }
+      Response response = ExceptionUtils.uncheckedFactory(responseQueue::take);
+      if (response.getFlags() != expectedResponse) {
+        throw new TransactionRejectedException(transactionId);
+      }
+      if (response.hasSequenceId()) {
+        Preconditions.checkState(
+            sequenceId == -1 || sequenceId == response.getSeqId(),
+            "Updated sequence ID. That is a bug in proxima's transactions.");
+        sequenceId = response.getSeqId();
+      }
+      state = State.Flags.OPEN;
+    }
+
+    public void commitWrite(List<StreamElement> outputs, CommitCallback callback)
+        throws TransactionRejectedException {
+
+      List<StreamElement> toWrite =
+          outputs.stream().map(this::injectSequenceId).collect(Collectors.toList());
+      List<KeyAttribute> keyAttributes =
+          toWrite.stream().map(KeyAttributes::ofStreamElement).collect(Collectors.toList());
+      manager.commit(transactionId, keyAttributes);
+      Response response = ExceptionUtils.uncheckedFactory(responseQueue::take);
+      if (response.getFlags() != Response.Flags.COMMITTED) {
+        if (response.getFlags() == Response.Flags.ABORTED) {
+          // already aborted
+          needsRollback = false;
+        }
+        throw new TransactionRejectedException(transactionId);
+      }
+      CommitCallback compositeCallback = CommitCallback.afterNumCommits(toWrite.size(), callback);
+      state = State.Flags.COMMITTED;
+      toWrite.forEach(w -> delegate.write(w, compositeCallback));
+    }
+
+    private void enqueueResponse(String responseId, Response response) {
+      ExceptionUtils.unchecked(() -> responseQueue.put(response));
+    }
+
+    private StreamElement injectSequenceId(StreamElement in) {
+
+      Preconditions.checkArgument(!in.isDeleteWildcard(), "Wildcard deletes not yet supported");
+
+      return StreamElement.upsert(
+          in.getEntityDescriptor(),
+          in.getAttributeDescriptor(),
+          sequenceId,
+          in.getKey(),
+          in.getAttribute(),
+          in.getStamp(),
+          in.getValue());
+    }
+
+    @Override
+    public void close() {
+      if (needsRollback && state != State.Flags.COMMITTED) {
+        rollback();
+      }
+    }
+
+    public void rollback() {
+      manager.rollback(transactionId);
+    }
+  }
+
+  private final OnlineAttributeWriter delegate;
+  private final ClientTransactionManager manager;
+
+  private TransactionalOnlineAttributeWriter(
+      DirectDataOperator direct, OnlineAttributeWriter delegate) {
+
+    this.delegate = delegate;
+    this.manager = TransactionResourceManager.of(direct);
+  }
+
+  @Override
+  public URI getUri() {
+    return delegate.getUri();
+  }
+
+  @Override
+  public void close() {
+    delegate.close();
+  }
+
+  @Override
+  public void write(StreamElement data, CommitCallback statusCallback) {
```

Comment:
As far as I understand, creating & committing transaction isn't cheap, so we may get a huge performance hit in this case.

But this is not intended for a highly stressed code paths, so it may be ok for now. In the future we may need to consider a variant that batches multiple writes for throughput.

Reply:
I think there is something that can be done better. The write is apparently asynchronous, so I'll change that inside the method to be asynchronous as well. Then I suppose the performance penalty will be "nearly" negligible - yes, it would have to hit several commit-log back and forth, but in terms of throughput it will be a constant factor.